adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1028 @@
+		return nil, nil
+	}
+
+	httpStatusError := checkRespStatusCode(bidderRawResponse)
+	if httpStatusError != nil {
+		return nil, []error{httpStatusError}
+	}
+
+	var huaweiAdsResponse HuaweiAdsResponse
+	if err := json.Unmarshal(bidderRawResponse.Body, &huaweiAdsResponse); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "Bad Server Response",
+		}}
+	}
+
+	huaweiAdsResponseError := checkHuaweiAdsResponseRetcode(huaweiAdsResponse)
+	if huaweiAdsResponseError != nil {
+		return nil, []error{huaweiAdsResponseError}
+	}
+
+	var err error
+	bidderResponse, err = convertHuaweiAdsResp2BidderResp(&huaweiAdsResponse, openRTBRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return bidderResponse, nil
+}
+
+// Builder builds a new instance of the HuaweiAds adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &HuaweiAdsAdapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// get request header
+func getHeaders(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, request *openrtb2.BidRequest, isAddAuthorization bool) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	if isAddAuthorization {
+		headers.Add("Authorization", getDigestAuthorization(huaweiAdsImpExt))
+	}
+
+	if request.Device != nil && len(request.Device.UA) > 0 {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+
+	return headers
+}
+
+// get body json for HuaweiAds request
+func getHuaweiAdsReqJson(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest, huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) error {
+	request.Version = HuaweiAdxApiVersion
+	var err error
+	if err = getHuaweiAdsReqAppInfo(request, openRTBRequest); err != nil {
+		return err
+	}
+	if err = getHuaweiAdsReqDeviceInfo(request, openRTBRequest, huaweiAdsImpExt); err != nil {
+		return err
+	}
+	getHuaweiAdsReqNetWorkInfo(request, openRTBRequest)
+	getHuaweiAdsReqRegsInfo(request, openRTBRequest)
+	getHuaweiAdsReqGeoInfo(request, openRTBRequest)
+	return nil
+}
+
+// get adslot30
+func getHuaweiAdsReqAdslot30(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds,
+	openRTBImp *openrtb2.Imp, openRTBRequest *openrtb2.BidRequest) (Adslot30, error) {
+	adtypeLower := strings.ToLower(huaweiAdsImpExt.Adtype)
+	var adslot30 = Adslot30{
+		Slotid: huaweiAdsImpExt.SlotId,
+		Adtype: convertAdtypeString2Integer(adtypeLower),
+		Test:   int32(openRTBRequest.Test),
+	}
+
+	if adtypeLower == "roll" {
+		if openRTBImp.Video != nil {
+			adslot30.TotalDuration = int32(openRTBImp.Video.MaxDuration)
+		} else {
+			return adslot30, errors.New("GetHuaweiAdsReqAdslot30: MaxDuration is empty when adtype is roll.")
+		}
+	}
+	return adslot30, nil
+}
+
+// convert adtype String -> Integer
+func convertAdtypeString2Integer(adtypeLower string) int32 {
```

Comment:
You have defined these numbers above as constants with corresponding names. Why are you not using those constants here?

Reply:
adtype is integer, but the user parameter adtype is string, I can only transfer it

Reply:
I don't think you understood my intent. Just below, you could do:
```
	if adtypeLower == "banner" {
		return banner
	} else if adtypeLower == "native" {
		return native
...
```
This makes the code a little cleaner I think, and insures you are at least consistent with the numerical value for "banner" and all.

Reply:
corrected